onnxscript/rewriter/__init__.py

```diff
@@ -5,53 +5,85 @@
+)
 
-PatternRewriteRule = pattern.RewriteRule
-
-ModelProtoOrIr = TypeVar("ModelProtoOrIr", onnx.ModelProto, ir.Model)
+_ModelProtoOrIr = TypeVar("_ModelProtoOrIr", onnx.ModelProto, ir.Model)
+_DEFAULT_REWRITE_RULES: tuple[pattern.RewriteRule, ...] = (
+    *no_op.rules.rules,  # TODO: merge this rule into constant folding?
+    *broadcast_to_matmul.rules.rules,
+    gemm_to_matmul_add.rule,  # type: ignore[has-type]
+    *cast_constant_of_shape.rules.rules,
+    *collapse_slices.rules.rules,
+    *llama_rule_sets.llama_p0_rule_set().rules,
+)
 
 
 class RewritePass(ir.passes.InPlacePass):
     def __init__(
         self,
-        pattern_rewrite_rules: Sequence[PatternRewriteRule] | pattern.RewriteRuleSet = (),
+        /,
+        rules: Sequence[pattern.RewriteRule] | pattern.RewriteRuleSet,
     ) -> None:
-        if pattern_rewrite_rules:
-            if not isinstance(pattern_rewrite_rules, pattern.RewriteRuleSet):
-                # Create a pattern rule-set using provided rules
-                pattern_rewrite_rules = pattern.RewriteRuleSet(pattern_rewrite_rules)
-        assert isinstance(pattern_rewrite_rules, pattern.RewriteRuleSet)
-        self.pattern_rewrite_rules: pattern.RewriteRuleSet = pattern_rewrite_rules
+        super().__init__()
+        if isinstance(rules, Sequence):
+            if not rules:
+                raise ValueError("rules must not be empty")
+            # Create a pattern rule-set using provided rules
+            rules = pattern.RewriteRuleSet(rules)
+        assert isinstance(rules, pattern.RewriteRuleSet)
+        self.rules: pattern.RewriteRuleSet = rules
 
     def call(self, model: ir.Model) -> ir.passes.PassResult:
-        count = self.pattern_rewrite_rules.apply_to_model(model)
+        count = self.rules.apply_to_model(model)
         if count:
             print(f"Applied {count} of general pattern rewrite rules.")
         return ir.passes.PassResult(model, bool(count))
 
 
 def rewrite(
-    model: ModelProtoOrIr,
-    pattern_rewrite_rules: Union[Sequence[PatternRewriteRule], pattern.RewriteRuleSet] = (),
-) -> ModelProtoOrIr:
+    model: _ModelProtoOrIr,
+    pattern_rewrite_rules: Union[Sequence[pattern.RewriteRule], pattern.RewriteRuleSet]
+    | None = None,
+) -> _ModelProtoOrIr:
+    """Rewrite the model using the provided pattern rewrite rules.
+
+    Unused nodes, functions, and opsets will be removed after the rewrite.
+
+    Args:
+        model: The model to be rewritten. Can be an ONNX ModelProto or an ir.Model.
+        pattern_rewrite_rules: A sequence of pattern rewrite rules or a RewriteRuleSet.
+            If not provided or empty, default rules will be applied.
+
+    Returns:
+        The rewritten model, either as an ONNX ModelProto or an ir.Model.
+    """
     if isinstance(model, onnx.ModelProto):
         model_ir = ir.serde.deserialize_model(model)
         proto = True
     else:
         model_ir = model
         proto = False
 
+    if not pattern_rewrite_rules:
```

Comment:
(Discussing a real edge case here): Should we distinguish between None and an empty list here? If the user does not specify an explicit rewrite-rule set (i.e., None), using the default rules makes sense. If the user explicitly specifies an empty list, ideally it should be treated as a no-op (OR be treated as an error, which is slightly less preferable)?

I suggest
```py
   if pattern_rewrite_rules is None:
      pattern_rewrite_rules = _DEFAULT_REWRITE_RULES
   else if pattern_rewrite_rules == []:
      return model
```

And suggest moving this to the beginning of the function (avoid unnecessary deserialization in the return-case).

Reply:
Sure! Done.